core/iwasm/aot/arch/aot_reloc_riscv.c

```diff
@@ -170,14 +170,32 @@ rv_get_val(uint16 *addr)
     return ret;
 }
 
-/* Set a val to given address */
+/**
+ * Set a val to given address
+ *
+ * Zifencei extension is moved out from I extension in RISC-V spec 20191213
+ * (rv32i/rv64i version 2.1 and later), new version of GNU toolchain or clang
+ * may not support it if no explicit 'zifencei' is specified in arch attribute.
+ *
+ * Refer to:
+ * https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md
+ */
+#if __riscv_i >= 2001000 && !defined(__ZEPHYR__)
+__attribute__((target("arch=+zifencei")))
+#endif
 static void
 rv_set_val(uint16 *addr, uint32 val)
 {
     *addr = (val & 0xffff);
     *(addr + 1) = (val >> 16);
 
+    /* REVISIT: Zephyr's GCC toolchain does not support attribute
+     * target("arch=+zifencei") so we must check if the extension is
+     * enabled at runtime.
+     */
```

Comment:
I don't think it is the case that the toolchain does not support the attribute. Rather, if a target does not support `zifencei` then the attribute will not be available. This is true whether using the Zephyr provided toolchain or another toolchain. Also, I don't think we are evaluating at runtime, the check performed below is evaluated at build time.

Reply:
I belive the toolchain support this attribute but don't support enable it by grammer `__attribute__(target("arch=+zifenci"))`, zephyr's toolchain will report:
```
/root/wasm-micro-runtime/core/iwasm/aot/arch/aot_reloc_riscv.c:188:1: warning: target attribute is not supported on this machine [-Wattributes]
  188 | {
      | ^
/root/wasm-micro-runtime/core/iwasm/aot/arch/aot_reloc_riscv.c: Assembler messages:
/root/wasm-micro-runtime/core/iwasm/aot/arch/aot_reloc_riscv.c:192: Error: unrecognized opcode `fence.i'
/root/wasm-micro-runtime/core/iwasm/aot/arch/aot_reloc_riscv.c:192: Error: unrecognized opcode `fence.i'
```
But it supported by other toolchain such as xpack's or official toolchain since this grammer is supported in spec:
https://github.com/riscv-non-isa/riscv-c-api-doc/blob/73877888d0d0e0a3a938b811d81a212a5edc4e3c/riscv-c-api.md?plain=1#L209

Reply:
Or the given target is not support the zifenci, could you share the `-march` parameter from zephyr?

Reply:
> Or the given target is really don't support the zifenci extension, could you share the -march parameter from zephyr?

This is correct -- the `esp32c3_devkitm` target does not support the `zifencei` extension, but others do, such as `qemu_riscv64`.